File: Backend/Services/Vacancy/Implementations/VacancyService.cs
using Backend.DTOs.Vacancy;
using Backend.Repositories.Vacancy.Interfaces;
using Backend.Services.Vacancy.Interfaces;

namespace Backend.Services.Vacancy.Implementations
{
    public class VacancyService : IVacancyService
    {
        private readonly IVacancyRepository _vacancyRepository;

        public VacancyService(
            IVacancyRepository vacancyRepository)
        {
            _vacancyRepository = vacancyRepository;
        }

        public async Task<Backend.Models.Vacancy.Vacancy>
            CreateAsync(CreateVacancyDto dto)
        {
            var vacancy =
                new Backend.Models.Vacancy.Vacancy
                {
                    CompanyId = dto.CompanyId,
                    JobTitle = dto.JobTitle,
                    JobDescription = dto.JobDescription,
                    RequiredSkills = dto.RequiredSkills,
                    RequiredExperience = dto.RequiredExperience,
                    Education = dto.Education,
                    Location = dto.Location,
                    EmploymentType = dto.EmploymentType,
                    ClosingDate = dto.ClosingDate,
                    Status = "Open",
                    CreatedAt = DateTime.UtcNow
                };

            return await _vacancyRepository
                .CreateAsync(vacancy);
        }

        public async Task<List<Backend.Models.Vacancy.Vacancy>>
            GetByCompanyIdAsync(int companyId)
        {
            return await _vacancyRepository
                .GetByCompanyIdAsync(companyId);
        }

        public async Task<Backend.Models.Vacancy.Vacancy?> GetByIdAsync(
    int vacancyId)
        {
            return await _vacancyRepository.GetByIdAsync(vacancyId);
        }

        public async Task<bool> UpdateAsync(
    int vacancyId,
    UpdateVacancyDto dto)
        {
            var vacancy = await _vacancyRepository
                .GetByIdAsync(vacancyId);

            if (vacancy == null)
            {
                return false;
            }

            vacancy.JobTitle = dto.JobTitle;
            vacancy.JobDescription = dto.JobDescription;
            vacancy.RequiredSkills = dto.RequiredSkills;
            vacancy.RequiredExperience = dto.RequiredExperience;
            vacancy.Education = dto.Education;
            vacancy.Location = dto.Location;
            vacancy.EmploymentType = dto.EmploymentType;
            vacancy.ClosingDate = dto.ClosingDate;
            vacancy.UpdatedAt = DateTime.UtcNow;

            return await _vacancyRepository.UpdateAsync(vacancy);
        }

        public async Task<bool> CloseAsync(int vacancyId)
        {
            var vacancy = await _vacancyRepository
                .GetByIdAsync(vacancyId);

            if (vacancy == null)
            {
                return false;
            }

            vacancy.Status = "Closed";
            vacancy.UpdatedAt = DateTime.UtcNow;

            return await _vacancyRepository.UpdateAsync(vacancy);
        }
    }
}

File: Frontend/src/app/core/services/vacancy.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CreateVacancyDto, UpdateVacancyDto, VacancyModel } from '../models/vacancy.model';

@Injectable({
  providedIn: 'root'
})
export class VacancyService {

  private apiUrl =
    'https://localhost:7182/api/employer/vacancy';

  private readonly seekerApiUrl = 'https://localhost:7182/api/employer/vacancy';

  constructor(
    private http: HttpClient
  ) {}

  getMyVacancies(
    companyId: number
  ): Observable<any[]> {

    return this.http.get<any[]>(
      `${this.apiUrl}/company/${companyId}`
    );
  }

  getVacancy(
    vacancyId: number
  ): Observable<any> {

    return this.http.get<any>(
      `${this.apiUrl}/${vacancyId}`
    );
  }
  // Separate Job Seeker method preserves the existing Employer methods and URL.
  getSeekerVacancy(vacancyId: number) {
    return this.http.get<{
      vacancyId: number; companyId: number; jobTitle: string; jobDescription: string;
      requiredSkills: string; requiredExperience: string | null; education: string | null;
      location: string | null; employmentType: string | null; closingDate: string | null;
      status: string; createdAt: string;
    }>(`${this.seekerApiUrl}/${vacancyId}`);
  }
  getOpenVacancies() {
    return this.http.get<{
      vacancyId: number; companyId: number; jobTitle: string; jobDescription: string;
      requiredSkills: string; requiredExperience: string | null; education: string | null;
      location: string | null; employmentType: string | null; closingDate: string | null;
      status: string; createdAt: string;
    }[]>(this.seekerApiUrl);
  }
  createVacancy(dto: CreateVacancyDto): Observable<VacancyModel | null> {
    return this.http.post<VacancyModel | null>(this.apiUrl, dto);
  }

  updateVacancy(vacancyId: number, dto: UpdateVacancyDto): Observable<{ message: string } | null> {
    return this.http.put<{ message: string } | null>(`${this.apiUrl}/${vacancyId}`, dto);
  }

  closeVacancy(vacancyId: number): Observable<{ message: string } | null> {
    return this.http.put<{ message: string } | null>(`${this.apiUrl}/${vacancyId}/close`, {});
  }
}
